src/js/common/validate.js

```diff
@@ -0,0 +1,343 @@
+
+If there are no validation errors, we return `undefined`.  If there are validation errors, the raw output generated by
+AJV is transformed using the static function `sanitizeValidationErrors`.
+
+As an example, let's assume that you have a JSON payload like the following that you wish to validate:
+
+  ```
+  {
+    field1: "invalid",
+    category1: {
+      nestedField1: "invalid"
+    }
+  }
+  ```
+
+Assuming that each value was invalid, you might see validation output like the following:
+
+  ```
+  {
+    fieldErrors: {
+      field1: ["should NOT be shorter than 8 characters", "should match pattern \"[A-Z]+\""],
+      category1: {
+        nestedField1: ["nested field is not a valid number."]
+      }
+  }
+  ```
+
+If you have configured your instance to work together with a `gpii.schema.parser`, it will further evolve the error
+messages where possible.
+
+ */
+
+"use strict";
+var fluid = fluid || require("infusion");
+var gpii  = fluid.registerNamespace("gpii");
+
+var Ajv = require("ajv");
+
+fluid.registerNamespace("gpii.schema.validator");
+
+gpii.schema.validator.init = function (that) {
+    // We persist a single AJV instance so that we can take advantage of its automatic compiling and caching.
+    that.ajv = Ajv(that.options.validatorOptions); // jshint ignore:line
+
+    gpii.schema.validator.refreshSchemas(that);
+};
+
+gpii.schema.validator.validate = function (that, key, content) {
+    var contentValid = that.ajv.validate(key, content);
+    if (!contentValid) {
+        return (gpii.schema.validator.sanitizeValidationErrors(that, key, that.ajv.errors));
+    }
+
+    return undefined;
+};
+
+
+/*
+ AJV gives us output like:
+
+ [
+     {
+         "keyword": "minLength",
+         "dataPath": ".password",
+         "message": "should NOT be shorter than 8 characters"
+     },
+     {
+         "keyword": "pattern",
+         "dataPath": ".password",
+         "message": "should match pattern \"[A-Z]+\""
+     },
+     {
+         "keyword": "required",
+         "dataPath": ".deep.required",
+         "message": "is a required property"
+     }
+ ]
+
+ We need to break this down so that the messages can be associated with individual model variables and associated form
+ elements.  This allows us to show overall in-context feedback to the user.
+
+ */
+gpii.schema.validator.sanitizeValidationErrors = function (that, schemaKey, errors) {
+    var sanitizedErrors = { fieldErrors: {}};
+
+    fluid.each(errors, function (error) {
+        // Errors are associated with the right field based on the `dataPath` received from AJV.
+        var path = gpii.schema.validator.extractPathSegmentsFromError(error);
+
+        var errorMessage      = error.message;
+        var overwriteExisting = false;
+
+        // If we have a parser, we will evolve the output if possible, and use that to replace every raw message for
+        // the same field.
+        //
+        if (that.parser) {
```

Comment:
The parser was added later in the process, it was convenient to have the ability to work without it while I was migrating.  I have cleaned this up and collapsed the `hasParser` grades into their respective defaults.